firedrake/ensemble/ensemble_function.py

```diff
@@ -0,0 +1,292 @@
+            One subset for each local :class:`Function`. None elements
+            will be ignored.  The values of each local function will
+            only be assigned on the nodes on the corresponding subset.
+        """
+        if type(other) is not type(self):
+            raise ValueError(
+                f"Cannot assign {type(self)} from {type(other)}")
+        for i in range(self._fs.nlocal_spaces):
+            self.subfunctions[i].assign(
+                other.subfunctions[i],
+                subset=subsets[i] if subsets else None)
+        return self
+
+    @PETSc.Log.EventDecorator()
+    def copy(self):
+        """
+        Return a deep copy of the :class:`EnsembleFunction`.
+        """
+        new = type(self)(self.function_space())
+        new.assign(self)
+        return new
+
+    @PETSc.Log.EventDecorator()
+    def zero(self, subsets=None):
+        """
+        Set values to zero.
+
+        Parameters
+        ----------
+
+        subsets : Collection[Optional[:class:`pyop2.types.set.Subset`]]
+            One subset for each local :class:`Function`. None elements
+            will be ignored.  The values of each local function will
+            only be zeroed on the nodes on the corresponding subset.
+        """
+        for i in range(self._fs.nlocal_spaces):
+            self.subfunctions[i].zero(
+                subset=subsets[i] if subsets else None)
+        return self
+
+    @PETSc.Log.EventDecorator()
+    def __iadd__(self, other):
+        for us, uo in zip(self.subfunctions, other.subfunctions):
+            us.assign(us + uo)
+        return self
+
+    @PETSc.Log.EventDecorator()
+    def __imul__(self, other):
+        if type(other) is type(self):
+            for us, uo in zip(self.subfunctions, other.subfunctions):
+                us.assign(us*uo)
+        else:
+            for us in self.subfunctions:
+                us *= other
+        return self
+
+    @PETSc.Log.EventDecorator()
+    def __add__(self, other):
+        new = self.copy()
+        new += other
+        return new
+
+    @PETSc.Log.EventDecorator()
+    def __mul__(self, other):
+        new = self.copy()
+        new *= other
+        return new
+
+    @PETSc.Log.EventDecorator()
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    @contextmanager
+    def vec(self):
+        """
+        Context manager for the global PETSc Vec with read/write access.
+
+        It is invalid to access the Vec outside of a context manager.
+        """
+        # _fbuf.vec shares the same storage as _vec, so we need this
+        # nested context manager so that the data gets copied to/from
+        # the Function.dat storage and _vec.
+        # However, this copy is done without _vec knowing, so we have
+        # to manually increment the state.
+        with self._fbuf.dat.vec:
+            self._vec.stateIncrease()
+            yield self._vec
```

Comment:
I don't really understand the need for this. Could you simply have
```py
def vec(self):
    return self._fbuf.dat.vec
```
or
```py
@contextlib.contextmanager
def vec(self):
    yield from self._fbuf.dat.vec
```
?

Reply:
See the implementation notes I've added to the PR description. Let me know if they don't make sense and/or you think there's a better way of doing it.